feat(processor): introduce internals flag

frisitano opened the pull request:
This PR introduces the `internals` feature flag

hackaugusto commented on processor/src/lib.rs:
```diff
@@ -443,4 +443,43 @@ where
+
+    /// Returns the process stack.
+    #[cfg(feature = "internals")]
+    pub fn stack(&self) -> &Stack {
```
ATM this is the only method used by the tests on `miden-base`. I see no harm in adding the other methods behind the flag though.

hackaugusto reviewed (approved):
LGTM, I'm not merging to wait for @bobbinth 's opinion.

vlopes11 commented on processor/src/lib.rs:
```diff
@@ -443,4 +443,43 @@ where
+
+    /// Returns the process system.
+    #[cfg(feature = "internals")]
+    pub fn system(&self) -> &System {
```
small nit: all these fns should be `const` so we can use them as static variables

vlopes11 reviewed (approved):
Looks good! However, none of these methods looks particularly harmful as they are just accessors.

Is there any reason why we won't just do the following?

```rust
let Process {
  decoder,
  range,
  .. } = process;
```

Also, maybe it would be good for the long run to add some comment on why we don't want to expose these via the standard API, as they pose no risk whatsoever to the internal correctness of the process.

hackaugusto commented:
> Looks good! However, none of these methods looks particularly harmful as they are just accessors.

Yes, @bobbinth wants to make `Processor` into a private API, so we are exposing this for tests only.

vlopes11 commented:
> > Looks good! However, none of these methods looks particularly harmful as they are just accessors.
> 
> Yes, @bobbinth wants to make `Processor` into a private API, so we are exposing this for tests only.
> 
> Edit: Or maybe I misunderstood your question, and you're proposing doing the same thing as [`into_parts`](https://github.com/0xPolygonMiden/miden-vm/blob/98dfb1b6b201fdd9eb2395c9196af2cf3b18fca2/processor/src/lib.rs#L436-L445). The issue there was that this consumes the `Processor`, and these methods don't

Could be something similar to into_parts, yes. Being just a test, we shouldn't really care if we have extra clones as performance is not a concern.

But it's just different ways to approach the same problem. I just wonder if we should create extra code for something that can be solved with the current code.

frisitano commented:
> Could be something similar to into_parts, yes. Being just a test, we shouldn't really care if we have extra clones as performance is not a concern.
> 
> But it's just different ways to approach the same problem. I just wonder if we should create extra code for something that can > be solved with the current code.

I didn't realise this could be solved in another way.  I think the point @vlopes11 makes is valid, we probably shouldn't be introducing this unless it's really necessary.  What do you think @hackaugusto?

hackaugusto commented:
The idea was not performance but readability. But I don't mind either way.

If the number of lines of code is a worry, then I suppose we could also have this struct in a private module and pub visibility:

```rust
#[cfg(feature = "internals")]
pub use internals::A;

mod internals {
    pub struct A {
        pub a: usize,
    }
}
```

That way we can access the attributes without consuming the data, which leaves it behind to use its methods, and we don't need any accessors.

bobbinth commented:
Looks good! Thank you! I do have a couple of questions:

1. How is this going to be consumed in the `miden` crate (or anywhere else)? (i.e., if we want to have a test which accesses internal functionality of `processor` crate).
2. Instead of marking each method as internal, could we make the entire `Process` struct internal? (i.e., something similar to what is suggested @hackaugusto in https://github.com/0xPolygonMiden/miden-vm/pull/714#issuecomment-1440638172. I've never felt great about having `Process` struct be a part of the public interface, and the only reason it is public now is for testing purposes.

frisitano commented:
Following @hackaugusto and @bobbinth suggestion I have modified the `Process` struct such that it is private by default.  If the `internals` flag is provided then the `Process` struct and it's internals are exposed publicly.  Let me know what you think of this pattern.  One nit is that we will now have to specify `--features "internals"` when intending to run the full test suite but I don't think this is too big of a deal.

hackaugusto commented on processor/src/lib.rs:
```diff
@@ -137,7 +137,8 @@ where
 // ================================================================================================
 
-pub struct Process<A>
+#[cfg(not(feature = "internals"))]
```
question: why not using the pattern with the private module and option pub use?

I'm asking because I assume that version would have less code duplication, namely the `Process` struct would be defined only once. And line of codes was one of the motivations for that suggestion.

frisitano replied:
I believe your proposal was to do something like the following.  I think this would work and is definitely a pattern that we can consider but I don't think it's as readable personally.  Happy to go with the consensus on this.

```rust
#[cfg(feature = "internals")]
pub use processor_internal::Process;

#[cfg(not(feature = "internals"))]
use processor_internal::Process;

mod processor_internal {
    use super::*;

    pub struct Process<A>
    where
        A: AdviceProvider,
    {
        pub system: System,
        pub decoder: Decoder,
        pub stack: Stack,
        pub range: RangeChecker,
        pub chiplets: Chiplets,
        pub advice_provider: A,
    }
}
```

hackaugusto commented:
thanks for taking care of this! I left one comment, but overall looks good to me :)

vlopes11 commented on .github/workflows/ci.yml:
```diff
@@ -50,7 +50,7 @@ jobs:
         with:
           command: test
-          args: ${{matrix.args}}
+          args: ${{matrix.args}} --features "internals"
```
optional nit that can be come an issue for the future: we should probably find an efficient way to permutate all the possible combinations of features.

vlopes11 commented on miden/Cargo.toml:
```diff
@@ -32,10 +32,16 @@ harness = false
 concurrent = ["prover/concurrent", "std"]
 default = ["std"]
-executable = ["env_logger", "hex/std", "std", "serde/std", "serde_derive", "serde_json/std", "structopt", "rustyline"]
+executable = ["env_logger", "hex/std", "internals", "std", "serde/std", "serde_derive", "serde_json/std", "structopt", "rustyline"]
```
It's not immediately obvious why we would need `executable`, that's the endpoint of the workspace, to depend on `internals`. I believe `internals` are meant only for non-production stuff.

Is there any way to circumvent this?

frisitano replied:
the [repl](https://github.com/0xPolygonMiden/miden-vm/blob/next/miden/src/repl/mod.rs#L273) has a dependency on the `Process` struct.  But I agree we should see if we can remove it.

bobbinth replied:
Let's create an issue for this so that we don't forget to fix it later.

vlopes11 commented on processor/src/chiplets/mod.rs:
```diff
@@ -29,7 +29,7 @@ use kernel_rom::KernelRom;
 pub use bus::{AuxTraceBuilder, ChipletsBus};
 
-#[cfg(test)]
+#[cfg(all(test, feature = "internals"))]
```
Shouldn't it be `any`? Meaning: It is going to be activated if either `test` "OR" `internals` are enabled. `all` will `test` "AND" `internals`

frisitano replied:
The tests module has a dependency on `internals` so we will always need `internals` active when `test` is. For the other case we could compile the `tests`  if only `internals` is active `test` is not but I don't think that makes sense.

frisitano replied:
Upon review I think you're right.  If I consistently mark as `any` it works as expected.  Good catch, I'll get this updated tomorrow.

vlopes11 commented on processor/src/decoder/aux_hints.rs:
```diff
@@ -316,7 +316,7 @@ impl BlockStackTableRow {
     }
 
-    #[cfg(test)]
+    #[cfg(all(test, feature = "internals"))]
```
Ditto; this looks like `any`

vlopes11 commented on processor/src/decoder/mod.rs:
```diff
@@ -28,7 +28,7 @@ pub use aux_hints::{
 };
 
-#[cfg(test)]
+#[cfg(all(test, feature = "internals"))]
```
Ditto; `any`

vlopes11 commented on processor/src/trace/mod.rs:
```diff
@@ -169,7 +169,7 @@ impl ExecutionTrace {
     }
 
-    #[cfg(test)]
+    #[cfg(feature = "internals")]
```
Shouldn't we follow the pattern of the others and activate this part if we have either test or the feature?

bobbinth replied:
We can set this to `#[cfg(any(test, feature = "internals"))]` - but only after `pub` version or process enabled for tests as well.

vlopes11 reviewed (commented):
Looks good overall; couple of small questions

bobbinth commented on processor/src/lib.rs:
```diff
@@ -444,3 +445,19 @@ where
+// INTERNALS
+// ================================================================================================
+
+#[cfg(feature = "internals")]
```
I would change this to `#[cfg(any(test, feature = "internals"))]` (and update line 140 accordingly). This way, we don't need to add any extra attributes to internal test modules.

bobbinth reviewed (changes requested):
Looks good! Thank you! I left a couple of small comments inline.

So, I guess now to run the tests as we do in the CI we need to run something like this:
```Shell
RUSTFLAGS="-C debug-assertions" cargo test --release --features internals
```
I'm wondering if there is a way to put this under a single command somehow.

frisitano commented:
So I've reviewed the feedback and I've converged on the following three rules regarding tagging testing objects:

1. If the object is only required for testing within the same crate it should be marked with `[cfg(test)]`
2. If the object is required for testing both within the same crate and some external crate it should be marked with `[cfg(any(test, feature = "internals"))]`
3 If the object is only required for testing of some external crate then it should be marked with `[cfg(feature = "internals")]`  (I'm not sure if we will ever experience this case).

I have updated the PR to reflect these rules and as such the only tag we need to add is `#[cfg(any(test, feature = "internals"))]` to the `Process` struct.  We then define the dependency of `miden/tests` on the `internals` flag in the `Cargo.toml` in the `miden` crate.  I think this is the cleanest approach.

I've also added a `Makefile` such that we can run the command:
```
RUSTFLAGS="-C debug-assertions" cargo test --release --features internals`
```
by executing `make test` from the root of the repo.

What do you guys think? @hackaugusto @bobbinth @vlopes11

frisitano commented:
> So, I guess now to run the tests as we do in the CI we need to run something like this:
> 
> ```shell
> RUSTFLAGS="-C debug-assertions" cargo test --release --features internals
> ```
> 
> I'm wondering if there is a way to put this under a single command somehow.

I've added a `Makefile` to the root of the repo.  See latest comment for my thoughts on the rest of the PR.

vlopes11 commented:
@frisitano that sounds very good! One small nit.

We probably want the command to be, instead:

```
RUSTFLAGS="-C debug-assertions -C overflow-checks -C debuginfo=2" cargo test --release --features internal
```

We should probably update the CI with that, too. Release mode will suppress overflow errors, and they are often very dangerous as they can easily create attack vectors.

vlopes11 commented on Makefile:
```diff
@@ -0,0 +1,2 @@
+test:
+	RUSTFLAGS="-C debug-assertions" cargo test --release --features internals
```
optional nit: line break @ EOF

frisitano replied:
Added new line and updated the command as you suggested.  Thank you!

vlopes11 reviewed (approved):
LGTM! Thanks

vlopes11 commented:
One alternative is to keep everything within the same block, with a heavy cost to readability

```rust
pub struct Foo {
    #[cfg(feature = "internals")]
    pub bar: String,

    #[cfg(not(feature = "internals"))]
    pub bar: String,
}
```

This would have the advantage to decrease chances that breaking changes in `Process` will be applied in only one of the implementations, but also it would make the struct quite unreadable. From my side, whatever of the two you prefer is fine by me!

frisitano commented:
> One alternative is to keep everything within the same block, with a heavy cost to readability
> 
> ```rust
> pub struct Foo {
>     #[cfg(feature = "internals")]
>     pub bar: String,
> 
>     #[cfg(not(feature = "internals"))]
>     bar: String,
> }
> ```
> 
> This would have the advantage to decrease chances that breaking changes in `Process` will be applied in only one of the implementations, but also it would make the struct quite unreadable. From my side, whatever of the two you prefer is fine by me!

I think the problem here is that we want the `Process` struct itself to be private.

frisitano commented:
> We probably want the command to be, instead:
> 
> ```
> RUSTFLAGS="-C debug-assertions -C overflow-checks -C debuginfo=2" cargo test --release --features internal
> ```
> 
> We should probably update the CI with that, too. Release mode will suppress overflow errors, and they are often very dangerous as they can easily create attack vectors.
> 
> I'm not very familiar with windows environments tho. Did Windows finally catch up or is it still unfriendly to use makefiles? I personally don't think we should heavily commit to windows requirements, but this is something to think about.
> 
> I did some quick research, and I saw a couple of suggestions to use `MinGW`, which is not very friendly/native. But then, windows users can just execute the command manually and it should be fine. Seems there is also a chocolatey package for make, so I guess that should suffice thinking

I have updated the command as you suggested.  Maybe we should create an issue to note that the CI should be updated?  I hadn't considered windows support for the `Makefile`.  As most of the team are using mac / linux I think we should be fine however I'm happy to change it.  Do you have any proposals?  Maybe a shell / bash script would be more portable?

vlopes11 commented:
@frisitano I think we can go without worrying about windows support too much.

If we are to support it, we can just recommend on the README to install `make` via the chocolatey package manager.

bobbinth reviewed (approved):
All looks good! Thank you!
